tamr_toolbox/utils/downstream.py

```diff
@@ -0,0 +1,245 @@
+        datasets_check = datasets_check.difference(downstream_datasets.keys()).difference(
+            {dataset.name}
+        )
+
+    # add the target dataset to list of downstream datasets to avoid checking later
+    downstream_datasets[dataset_name] = dataset
+
+    # check further downstream for either dependencies suggested by name
+    # or dependencies that used to or is unified dataset
+    for d in datasets_check:
+        _find_downstream_datasets(
+            d,
+            include_dependencies_by_name=include_dependencies_by_name,
+            client=client,
+            all_tamr_datasets=all_tamr_datasets,
+            all_unified_datasets=all_unified_datasets,
+            downstream_datasets=downstream_datasets,
+        )
+
+    return downstream_datasets
+
+
+def _find_associated_projects(dataset: Dataset) -> List[str]:
+    """Return list of project_list that the dataset is part of.
+
+    Args:
+        dataset: The target dataset.
+
+    Returns:
+        List of Project IDs that the target `dataset` is part of.
+
+    """
+    project_list = set(
+        [step.project().resource_id for step in dataset.usage().usage.input_to_project_steps]
+    )
+    return list(project_list)
+
+
+def _count_dependencies(dataset: Dataset) -> int:
+    """Return number of dependencies of a dataset.
+
+    Args:
+        dataset: The target dataset.
+
+    Returns:
+        Number of dependencies of the target `dataset`.
+
+    """
+    dependencies = dataset.usage().dependencies
+    return len(dependencies)
+
+
+def _is_unified_dataset(
+    dataset_name: str,
+    *,
+    check_regex: bool = True,
+    regex=".*unified_dataset$",
+    all_unified_datasets: Optional[List[str]] = None,
+) -> bool:
+    """Check if a dataset is or used to be unified dataset.
+
+    Args:
+        dataset: The target dataset.
+        check_regex: A Boolean to determine whether to check for regex matching if a dataset is
+            not part of the pipeline (if its project has been deleted.) Default to True.
+        regex: A string of regex to check. Default to '.*unified_dataset$' which is Tamr's current
+            default unified dataset naming.
+
+    Returns:
+        A Boolean, if True, the target `dataset` is or used to be unified dataset.
+
+    """
+    # check project pipeline (only works if project has not been deleted)
+    if dataset_name in all_unified_datasets:
+        return True
+    # check regex
+    if check_regex:
+        return re.search(regex, dataset_name) is not None
+    return False
+
+
+def _get_all_unified_datasets(client: Client) -> List[str]:
+    # Use Project Spec to minimize number of API calls
+    unified_dataset_names = [
+        project.spec().to_dict()["unifiedDatasetName"]
+        for project in client.projects
+        # GOLDEN_RECORDS project does not have unified dataset
+        if ProjectType[project.type] not in [ProjectType.GOLDEN_RECORDS]
```

Comment:
Lets make this opt-in not opt-out. So it specifically checks for mastering, schema-mapping, and categorization projects. This way if Tamr introduces more project types without UDs, this will not fail